realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp

```diff
@@ -0,0 +1,351 @@
+    } CATCH_STD()
+    return static_cast<jboolean>(false);
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_SharedRealm_nativeRefresh__J(JNIEnv *env, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    try {
+        shared_realm->refresh();
+    } CATCH_STD()
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_SharedRealm_nativeRefresh__JJJ(JNIEnv *env, jclass, jlong shared_realm_ptr,
+                                                      jlong version, jlong index) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    SharedGroup::VersionID version_id(static_cast<SharedGroup::version_type>(version),
+                                     static_cast<uint32_t>(index));
+    try {
+        using rf = realm::_impl::RealmFriend;
+        auto& shared_group = rf::get_shared_group(*shared_realm);
+        LangBindHelper::advance_read(shared_group, version_id);
+    } CATCH_STD()
+}
+
+JNIEXPORT jlongArray JNICALL
+Java_io_realm_internal_SharedRealm_nativeGetVersionID(JNIEnv *env, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    try {
+        using rf = realm::_impl::RealmFriend;
+        SharedGroup::VersionID version_id = rf::get_shared_group(*shared_realm).get_version_of_current_transaction();
+
+        jlong version_array [2];
+        version_array[0] = static_cast<jlong>(version_id.version);
+        version_array[1] = static_cast<jlong>(version_id.index);
+
+        jlongArray version_data = env->NewLongArray(2);
+        if (version_data == NULL) {
+            ThrowException(env, OutOfMemory, "Could not allocate memory to return versionID.");
+            return NULL;
+        }
+        env->SetLongArrayRegion(version_data, 0, 2, version_array);
+
+        return version_data;
+    } CATCH_STD ()
+
+    return NULL;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_SharedRealm_nativeIsClosed(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    return static_cast<jboolean>(shared_realm->is_closed());
+}
+
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_SharedRealm_nativeGetTable(JNIEnv *env, jclass, jlong shared_realm_ptr,
+                                                  jstring table_name) {
+    TR_ENTER_PTR(shared_realm_ptr)
+
+    try {
+        JStringAccessor name(env, table_name); // throws
+        auto shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+        if (!shared_realm->read_group().has_table(name) && !shared_realm->is_in_transaction()) {
+            std::ostringstream ss;
+            ss << "Table " << name << " doesn't exist and the shared realm is not in transaction.";
```

Comment:
Maybe also remove "shared" from the text as you mentioned the concept should not be exposed to users :-)